### PR TITLE
add shared <Link> component that uses react-router <Link> or <a>

### PR DESCRIPTION
**Pull request by sqs**
- If a component in `shared/` uses `react-router-dom`'s `<Link>` or `<NavLink>`, it will throw an error in the browser extension because the browser extension does not have an ancestor `react-router-dom` `<BrowserRouter>`.
- If a component in `shared/` uses `<a>` (and the `href` is a relative URL), it will likely have bad behavior: it will cause a full page reload instead of using HTML history API navigation (pushState).

This adds a new `<Link>` component in `shared/` that works correctly in both cases. This component should be used in all `shared/` code that needs a link.

It also adds a tslint rule to prevent imports of `react-router` or `react-router-dom` in `shared/` (which are unnecessary and will lead to these kinds of bugs).

See the docstring for the new `<Link>` component in `shared/src/components/Link.tsx` for more info.

### CODE REVIEW

File: shared/src/components/Link.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * react-router-dom <Link> wouldn't work in the browser extension, because there is no <BrowserRouter>.
+ *
+ * This variable must be set at initialization time by calling {@link setLinkComponent}.
+ *
+ * The `to` property holds the destination URL (do not use `href`). If <a> is used, the `to` property value is
+ * given as the `href` property value on the <a> element.
+ *
+ * @see setLinkComponent
+ */
+export let Link: React.ComponentType<
```
Comment (felixfbecker) [inline shared/src/components/Link.tsx]: Could this check for the presence of a router through `context` instead of using a global variable?

**Comment by vanesa** [comment]
I think I'm encountering the opposite issue of what you are trying to solve: https://github.com/sourcegraph/sourcegraph/issues/1365 
So in my case, the `<Link />` component is always rendering as a relative path, when an extension needs the link to be absolute.

**Comment by sqs** [comment]
@vanesa In your case, you should either just use `<a>` (or `<RouterLinkOrAnchor>`, which uses `<Link>` for relative URLs and `<a>` for absolute URLs). LMK if that helps.

**Comment by sqs** [comment]
@vanesa Oh, I see, the extension is sending the URL and you aren't using React directly. Yep, this is a separate bug (that I thought was fixed but apparently not).

File: shared/src/components/Link.tsx (hunk 1 above)
Comment (sqs) [reply shared/src/components/Link.tsx]: Whoops, thought I replied to this earlier. Yeah, let me try it that way.

Comment (sqs) [reply shared/src/components/Link.tsx]: This alternative impl is not possible.

The router-context appears to be fully encapsulated and not accessible outside of the package (without hacks).

There are no exports in the public API of it at https://sourcegraph.com/github.com/ReactTraining/react-router@master/-/blob/packages/react-router/modules/RouterContext.js#L12:16&tab=references.

If I try to import it from `react-router/es/RouterContext`, webpack fails:

```
01:12:33                    web | WARNING in ../shared/src/components/Link.tsx 36:49-62
01:12:33                    web | "export 'context' (imported as 'RouterContext') was not found in 'react-router/es/RouterContext'
ERROR in ../node_modules/react-router/es/RouterContext.js
01:12:33                    web | Module not found: Error: Can't resolve 'create-react-context' in '/home/sqs/src/github.com/sourcegraph/sourcegraph/node_modules/react-router/es'
```

**Pull request merged**